sycl/doc/extensions/proposed/sycl_ext_oneapi_memcpy2D.asciidoc

```diff
@@ -0,0 +1,137 @@
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Specification
+
+This extension adds support for 2D memcpy routines that can copy a specified
+rectangular region in the presence of array padding.
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_MEMCPY2D` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API Additions
+
+
+The handler class gains one new method:
+
+[cols="^60a,40"]
+|===
+| Member Function | Description
+
+a|
+[source,c++]
+----
+void ext_oneapi_memcpy2d(void *dest, size_t destPitch,
+  const void *src, size_t srcPitch,
+  size_t width, size_t height)
+----
+
+| Copies the data located at `src` to `dest`. At least one of `src` or `dest`
+must be a USM pointer. The width of a row in bytes,
+including any padding, is specified by `destPitch` for `dest` and `srcPitch`
+for `src`. The amount of data to copy is specified as `width` bytes per row
+times `height` number of rows.  `width` must not be larger than either
+`srcPitch` or `destPitch`, otherwise this function throws a synchronous
+`exception` with the `errc::invalid` error code.
+
+|===
+
+This extension also defines additional shortcut methods. The `queue` class
+gains three new methods:
+
+[cols="^60a,40"]
+|===
+| Member Function | Description
+
+|
+[source,c++]
+----
+event ext_oneapi_memcpy2d(void *dest, size_t destPitch,
+  const void *src, size_t srcPitch,
+  size_t width, size_t height)
+
+event ext_oneapi_memcpy2d(void *dest, size_t destPitch,
+  const void *src, size_t srcPitch,
+  size_t width, size_t height,
+  event depEvent)
+
+event ext_oneapi_memcpy2d(void *dest, size_t destPitch,
+  const void *src, size_t srcPitch,
+  size_t width, size_t height,
+  const std::vector<event> &depEvents)
+----
+
+| Copies the data located at `src` to `dest`. At least one of `src` or `dest`
+must be a USM pointer. The width of a row in bytes,
+including any padding, is specified by `destPitch` for `dest` and `srcPitch`
+for `src`.  The amount of data to copy is specified as `width` bytes per row
+times `height` number of rows.  `width` must not be larger than either
+`srcPitch` or `destPitch`, otherwise this function throws a synchronous
+`exception` with the `errc::invalid` error code.  If specified, the operation
```

Comment:
The core SYCL spec doesn't repeat all this information for the shortcut functions.  Instead, it just says:

>Equivalent to submitting a command group containing `handler::ext_oneapi_memcpy2d(dest, destPitch, src, srcPitch, width, height)`.